contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,478 @@
+        transformComponentInfo[_transformComponent] = _transformInfo;
+    }
+
+    /**
+     * ONLY OPERATOR: Updates TransformInfo entry for a transform component. Must already have been set before updating. If not,
+     * use setTransformInfo.
+     *
+     * @param _transformComponent       component that needs to be transformed/untransformed during rebalance
+     * @param _transformInfo            new TransfomInfo entry for the transform component
+     */
+    function updateTransformInfo(address _transformComponent, TransformInfo memory _transformInfo) external onlyOperator {
+        require(
+            transformComponentInfo[_transformComponent].underlyingComponent != address(0),
+            "TransformInfo not set yet"
+        );
+
+        require(_transformInfo.maxTransformSize <= type(uint96).max, "max transform size must be less than MAX_UINT_96");
+        transformComponentInfo[_transformComponent] = _transformInfo;
+    }
+
+    /**
+     * ONLY OPERATOR: Begins a rebalance. Must supply the set components as well as the target unit allocation for each. If the
+     * component is a transform component, then the target units should be measured in the equivalent value of the underlying tokens
+     *
+     * @param _setComponents            array of components involved in rebalance including components being removed (target units set to 0)
+     * @param _targetUnitsUnderlying    array of target units at end of rebalance, maps to same index of _components array
+     */
+    function startIPRebalance(address[] memory _setComponents, uint256[] memory _targetUnitsUnderlying) external onlyOperator {
+        require(_setComponents.length == _targetUnitsUnderlying.length, "length mismatch");
+        tradesComplete = false;
+
+        // clear out any current member of rebalanceParams from last rebalance
+        for (uint256 i = 0; i < setComponentList.length; i++) {
+            rebalanceParams[setComponentList[i]] = 0;
+        }
+
+        // Save rebalanceParams
+        for (uint256 i = 0; i < _setComponents.length; i++) {
+            rebalanceParams[_setComponents[i]] = _targetUnitsUnderlying[i];
+        }
+
+        setComponentList = _setComponents;
+    }
+
+    /**
+     * ONLY ALLOWED CALLER: Untransforms components. This function must be called after starting the rebalance, but before calling startTrades.
+     * The untransformData parameter can be fetched by the rebalance bots by calling getUntransformData on the relevent TransformHelper.
+     *
+     * @param _transformComponents      array of components to untransform
+     * @param _untransformData          array of untransform data 
+     */
+    function batchUntransform(
+        address[] memory _transformComponents,
+        bytes[] memory _untransformData
+    )
+        external
+        onlyAllowedCaller(msg.sender)
+    {
+        require(_transformComponents.length == _untransformData.length, "length mismatch");
```

Comment:
Is there a way we can try to enforce `This function must be called after starting the rebalance, but before calling startTrades`?

Reply:
While it is not explicitly enforced, assuming all the untransformations were called during the untransform step, it will actually be impossible to call untransform again at a later point in the rebalancing process. This is because this call would be guaranteed to always correctly calculate that the amount to transform is 0, which causes it to revert with "nothing to untransform".

I could add an enum that represents the current stage of the rebalance process we are in and keep updating it as we go along. This would allow us to then directly enforce that things are being done at the correct step.